File: backend/src/services/ai.service.js
const DEFAULT_AI_URL = "http://localhost:8000";

export class AiService {
  constructor() {
    this.baseUrl = process.env.AI_SERVICE_URL || DEFAULT_AI_URL;
    this.apiKey = process.env.AI_API_KEY || "";
  }

  headers() {
    const h = { "Content-Type": "application/json" };
    if (this.apiKey) {
      h["X-API-Key"] = this.apiKey;
    }
    return h;
  }

  async _parseJson(res) {
    const json = await res.json();
    if (!res.ok) {
      throw new Error(json.error || json.detail || `AI error ${res.status}`);
    }
    if (json.success === false) {
      throw new Error(json.error || "Error en servicio AI");
    }
    return json.data ?? json;
  }

  /** Predicción de rendimiento (nuevo esquema CosechaCoin). */
  async predictYield(payload) {
    const res = await fetch(`${this.baseUrl}/predict/yield`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(payload),
    });
    return this._parseJson(res);
  }

  /** Análisis NDVI con report_hash para complete_booster. */
  async analyzeDrone({ crop_type, region, booster_id }) {
    const res = await fetch(`${this.baseUrl}/drone/analyze`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ crop_type, region, booster_id }),
    });
    return this._parseJson(res);
  }

  /** Legacy: análisis por parcel_id (backend drone.service). */
  async analyzeNdvi(payload) {
    const res = await fetch(`${this.baseUrl}/analyze/ndvi`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`AI NDVI error ${res.status}: ${text}`);
    }
    return res.json();
  }

  async health() {
    const res = await fetch(`${this.baseUrl}/health`, {
      headers: this.headers(),
    });
    return res.json();
  }
}

export const aiService = new AiService();

File: mobile/lib/services/ai_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/constants.dart';

/// Cliente para el microservicio IA FastAPI (puerto 8000).
class AiService {
  AiService({String? baseUrl}) : _base = baseUrl ?? aiBaseUrl;

  final String _base;

  Map<String, String> get _headers {
    final h = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    if (aiApiKey.isNotEmpty) {
      h['X-Api-Key'] = aiApiKey;
    }
    return h;
  }

  /// POST /predict/yield
  Future<Map<String, dynamic>> predictYield({
    required String cropType,
    required String region,
    required double areaHectares,
  }) async {
    final res = await http.post(
      Uri.parse('$_base/predict/yield'),
      headers: _headers,
      body: jsonEncode({
        'crop_type': cropType,
        'region': region,
        'planted_area_hectares': areaHectares,
      }),
    );
    return _parseAiResponse(res);
  }

  /// POST /drone/analyze — reporte NDVI.
  Future<Map<String, dynamic>> analyzeDrone({
    required String cropType,
    required String region,
    int? boosterId,
  }) async {
    final res = await http.post(
      Uri.parse('$_base/drone/analyze'),
      headers: _headers,
      body: jsonEncode({
        'crop_type': cropType,
        'region': region,
        if (boosterId != null) 'booster_id': boosterId,
      }),
    );
    return _parseAiResponse(res);
  }

  Map<String, dynamic> _parseAiResponse(http.Response res) {
    Map<String, dynamic> body;
    try {
      body = jsonDecode(res.body) as Map<String, dynamic>;
    } catch (_) {
      throw AiException(
        'Respuesta inválida del servicio IA (${res.statusCode})',
      );
    }

    if (res.statusCode >= 200 && res.statusCode < 300) {
      if (body['success'] == false) {
        throw AiException(
          body['error']?.toString() ?? 'Error en predicción IA',
        );
      }
      final data = body['data'];
      if (data is Map<String, dynamic>) return data;
      return body;
    }

    final detail = body['detail'] ?? body['error'] ?? res.body;
    throw AiException(detail.toString());
  }
}

class AiException implements Exception {
  AiException(this.message);
  final String message;

  @override
  String toString() => message;
}
